### PR TITLE
[Integration][Custom] add custom authentication method

### PR DESCRIPTION
**Pull request by shalev007**
# Custom Authentication for HTTP Server Integration

This PR adds support for **custom authentication methods** to the `@integrations/custom` integration, enabling OAuth2, JWT, and other dynamic token-based authentication flows beyond the standard types (Bearer, Basic, API Key).

## Overview

When `authType` is set to `"custom"`, you configure:
1. **How to authenticate** (`customAuthRequest`) - defines the auth endpoint, method, headers, and body
2. **How to use the token** (`customAuthResponse`) - defines where to inject auth values in subsequent requests using template syntax

The integration automatically:
- Authenticates on startup
- Caches evaluated templates for performance
- Refreshes tokens proactively (if `reauthenticateIntervalSeconds` is configured) or reactively on 401 errors
- Handles concurrent requests safely with locking

### `customAuthRequest` (Required)

Defines how to make the authentication request:

| Field | Type | Description | Example |
|-------|------|-------------|---------|
| `endpoint` | string | Auth endpoint (relative path or full URL) | `"/oauth/token"` or `"https://auth.example.com/token"` |
| `method` | string | HTTP method (default: `"POST"`) | `"POST"`, `"GET"` |
| `headers` | object | HTTP headers for auth request | `{"Content-Type": "application/json"}` |
| `body` | object | JSON request body (mutually exclusive with `bodyForm`) | `{"grant_type": "client_credentials"}` |
| `bodyForm` | string | Form-encoded body (mutually exclusive with `body`) | `"grant_type=client_credentials&client_id=..."` |
| `queryParams` | object | Query parameters | `{"scope": "read"}` |
| `reauthenticateIntervalSeconds` | integer | Token validity duration in seconds. If not provided, tokens only refresh on 401 errors. A 60-second buffer refreshes proactively before expiration. | `3600` |

### `customAuthResponse` (Required)

Defines how to use the authentication response in subsequent API requests using template syntax `{{.jq_path}}`:

| Field | Type | Description | Example |
|-------|------|-------------|---------|
| `headers` | object | Headers to add/override with template values | `{"Authorization": "Bearer {{.access_token}}"}` |
| `queryParams` | object | Query params to add/override with template values | `{"api_key": "{{.access_token}}"}` |
| `body` | object | Body fields to add/override with template values | `{"token": "{{.accessToken}}"}` |

**Template Syntax:** Use `{{.jq_path}}` to extract values from the auth response. Supports nested paths like `{{.data.token}}` or `{{.nested.field.value}}`.

**Note:** At least one of `headers`, `queryParams`, or `body` must be provided.

### OAuth2 Client Credentials Flow

```yaml
authType: "custom"
customAuthRequest:
  endpoint: "/oauth/token"
  method: "POST"
  headers:
    Content-Type: "application/x-www-form-urlencoded"
  bodyForm: "grant_type=client_credentials&client_id=${CLIENT_ID}&client_secret=${CLIENT_SECRET}"
  reauthenticateIntervalSeconds: 3600  # Refresh every hour
customAuthResponse:
  headers:
    Authorization: "Bearer {{.access_token}}"
```

### API Key in Query Parameter

```yaml
authType: "custom"
customAuthRequest:
  endpoint: "/api/auth/login"
  method: "POST"
  headers:
    Content-Type: "application/json"
  body:
    username: "${USERNAME}"
    password: "${PASSWORD}"
customAuthResponse:
  queryParams:
    api_key: "{{.apiKey}}"
```

### Token in Multiple Places

```yaml
authType: "custom"
customAuthRequest:
  endpoint: "/v1/auth/token"
  method: "POST"
  body:
    clientId: "${CLIENT_ID}"
    clientSecret: "${CLIENT_SECRET}"
customAuthResponse:
  headers:
    Authorization: "Bearer {{.access_token}}"
    X-API-Version: "{{.version}}"
  queryParams:
    token: "{{.access_token}}"
  body:
    auth_token: "{{.access_token}}"
```

## How It Works

1. **Startup**: Integration authenticates once using `customAuthRequest` configuration
2. **Template Evaluation**: Values from auth response are extracted using JQ paths (e.g., `{{.access_token}}`)
3. **Request Injection**: Evaluated templates are injected into headers, query params, or body for each API request
4. **Token Refresh**: 
   - **Proactive**: If `reauthenticateIntervalSeconds` is set, tokens refresh 60 seconds before expiration
   - **Reactive**: On 401 errors, tokens automatically refresh and the request retries
5. **Caching**: Evaluated templates are cached until auth response changes, improving performance

## Migration

- **No Breaking Changes**: Existing auth types (`bearer_token`, `basic`, `api_key`, `none`) continue to work
- **New Feature**: Custom auth is opt-in - set `authType: "custom"` to use it

## Technical Details

- Template syntax is validated at configuration time (fail-fast)
- Thread-safe re-authentication prevents concurrent refresh attempts
- Shared client singleton ensures consistent auth state across all requests
- Supports environment variable substitution in config values (e.g., `${CLIENT_ID}`)

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Enhancement

___

### **Description**
- Add custom authentication handler with dynamic token retrieval and template evaluation

- Implement automatic re-authentication on 401 errors with module-level locking

- Support template-based auth value injection into headers, query params, and body

- Add shared client singleton pattern for parallel-safe operation across resync tasks

- Extend pagination handlers to support request body parameter passing

___

### Diagram Walkthrough

```mermaid
flowchart LR
  Config["Custom Auth Config<br/>customAuthRequest<br/>customAuthResponse"]
  Init["initialize_and_authenticate<br/>@ocean.on_start"]
  CustomAuth["CustomAuth Handler<br/>authenticate_async<br/>apply_auth_to_request<br/>reauthenticate"]
  Templates["Template Evaluation<br/>{{.jq_path}} syntax<br/>JQ expression support"]
  Client["HttpServerClient<br/>_make_request<br/>401 retry logic"]
  Request["API Request<br/>with auth headers<br/>query params<br/>body"]
  
  Config -->|parsed| Init
  Init -->|creates| CustomAuth
  CustomAuth -->|evaluates| Templates
  Templates -->|applies to| Client
  Client -->|on 401| CustomAuth
  CustomAuth -->|re-auth| Client
  Client -->|sends| Request
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>overrides.py</strong><dd><code>Add custom auth configuration models with validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/http_server/overrides.py

<ul><li>Add <code>CustomAuthRequestConfig</code> model with endpoint, method, headers, <br>body, bodyForm, and queryParams fields<br> <li> Add <code>CustomAuthResponseConfig</code> model with headers, queryParams, and body <br>fields for template-based auth value injection<br> <li> Implement root validators for mutual exclusivity of body/bodyForm and <br>HTTP method validation<br> <li> Support both dict and JSON string formats for configuration parsing</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-6003361257cf4cce4e0791c02f29c8980932078e3310cdaf5b72fdffc9351909">+76/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>handlers.py</strong><dd><code>Implement custom auth handler with template evaluation</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/http_server/handlers.py

<ul><li>Add <code>_evaluate_template()</code> async function to evaluate <code>{{.jq_path}}</code> <br>templates using Ocean's JQ processor<br> <li> Add <code>_evaluate_templates_in_dict()</code> for recursive template evaluation in <br>nested structures<br> <li> Implement <code>CustomAuth</code> class with <code>authenticate_async()</code>, <br><code>apply_auth_to_request()</code>, and <code>reauthenticate()</code> methods<br> <li> Add module-level <code>_reauthenticate_lock</code> to prevent concurrent <br>re-authentication race conditions<br> <li> Update all pagination handlers to accept optional <code>body</code> parameter in <br><code>fetch_all()</code> method<br> <li> Register <code>CustomAuth</code> in <code>AUTH_HANDLERS</code> registry and update <br><code>get_auth_handler()</code> factory</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-02dc3c095592539a02313c909d4101d589faf4ba1be99db18f8bae910329a40b">+345/-8</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>client.py</strong><dd><code>Add 401 handling and body parameter support</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/http_server/client.py

<ul><li>Add <code>custom_auth_request</code> and <code>custom_auth_response</code> parameters to <br><code>HttpServerClient.__init__()</code><br> <li> Add <code>skip_setup</code> parameter to defer authentication to <code>@ocean.on_start()</code> <br>hook<br> <li> Implement 401 re-authentication logic in <code>_make_request()</code> with <br>automatic token refresh<br> <li> Add request body support throughout fetch pipeline <br>(<code>fetch_paginated_data()</code>, <code>_fetch_with_pagination()</code>, <code>_make_request()</code>)<br> <li> Apply custom auth templates to headers, query params, and body before <br>each request<br> <li> Implement retry loop with max 1 retry on 401 before raising error</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-2b82e0d6b02d32f4e41a6b3fff5bf55aae7911b7da426a8e675d1a54e3820766">+95/-21</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>initialize_client.py</strong><dd><code>Add shared client singleton with async initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/initialize_client.py

<ul><li>Add module-level shared client singleton <code>_shared_client</code> with <br>authentication state tracking<br> <li> Implement <code>get_client()</code> async function for safe client access with <br>authentication wait logic<br> <li> Implement <code>initialize_and_authenticate()</code> async function for one-time <br>client initialization and auth<br> <li> Add parsing and validation of <code>customAuthRequest</code> and <code>customAuthResponse</code> <br>configurations<br> <li> Support both dict and JSON string formats for custom auth config <br>parsing<br> <li> Implement async locks and event flags to prevent race conditions in <br>parallel resync operations</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-511fea9e90c6ad3de90c59bf89f6c47ce21158b3cb915d212f5d9e2b1a58bd47">+166/-2</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Add on_start hook and update resync to use shared client</code>&nbsp; </dd></summary>
<hr>

integrations/custom/main.py

<ul><li>Add <code>@ocean.on_start()</code> hook to initialize shared client and <br>authenticate before resync operations<br> <li> Update <code>resync_resources()</code> to use <code>await get_client()</code> instead of <br><code>init_client()</code><br> <li> Add <code>body</code> parameter extraction from selector and pass through to <br><code>fetch_paginated_data()</code></ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-0fcca960cb3550c61820c8ad38abacd52cad18878ca87b24f32039eb331c0a03">+14/-3</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>spec.yaml</strong><dd><code>Add custom auth configuration to spec</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/.port/spec.yaml

<ul><li>Add <code>custom</code> to <code>authType</code> enum alongside existing auth methods<br> <li> Add <code>customAuthRequest</code> configuration field with dependency on <code>authType: </code><br><code>custom</code><br> <li> Add <code>customAuthResponse</code> configuration field with dependency on <br><code>authType: custom</code><br> <li> Update <code>authType</code> description to include custom authentication option</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-c428ac234d25ce406220e723ad1ff477fe4ccfedeae9670a3662a8af37b3779d">+22/-2</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>config-custom-auth.env</strong><dd><code>Add custom auth example configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/examples/config-custom-auth.env

<ul><li>Add example environment configuration file demonstrating custom <br>authentication setup<br> <li> Show <code>CUSTOM_AUTH_REQUEST</code> with endpoint, headers, and body for Port API <br>authentication<br> <li> Show <code>CUSTOM_AUTH_RESPONSE</code> with template syntax for Bearer token <br>injection<br> <li> Include commented pagination configuration options</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-1f1346d0a380e77a21eee32e6a60cbd08670a531b3ef5d0489768cad947a2633">+28/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_custom_auth.py</strong><dd><code>Add comprehensive custom auth test suite</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/tests/test_custom_auth.py

<ul><li>Add comprehensive test suite for template evaluation functions with JQ <br>path support<br> <li> Add <code>TestCustomAuth</code> class testing authentication flow, token storage, <br>and error handling<br> <li> Add <code>TestReauthentication</code> class testing 401 handling, lock behavior, <br>and concurrent auth prevention<br> <li> Add <code>TestClient401Handling</code> class testing end-to-end 401 retry and <br>re-authentication in client<br> <li> Include 690 lines of tests covering success paths, error cases, and <br>edge conditions</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2667/files#diff-7db496f340c4a32059c69bca67122d5b2e6daf4b6a2dfc399adaf18b5746291f">+690/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

### CODE REVIEW

File: integrations/custom/.port/spec.yaml [hunk 1]
```diff
@@ -156,6 +156,26 @@ configurations:
     type: string
     default: "10"
     description: "Maximum concurrent API requests (uses Ocean's built-in rate limiting and caching)"
+  - name: customAuthRequest
+    required: false
+    type: object
+    description: "Custom authentication request configuration. Required when authType is 'custom'. Defines how to make the authentication request. Supports optional 'reauthenticateIntervalSeconds' field to control token refresh timing. If not provided, expiration checking is disabled and tokens will only be refreshed on 401 errors. Example: {\"endpoint\": \"/oauth/token\", \"method\": \"POST\", \"headers\": {\"Content-Type\": \"application/json\"}, \"body\": {\"grant_type\": \"client_credentials\", \"client_id\": \"...\", \"client_secret\": \"...\"}, \"queryParams\": {}, \"reauthenticateIntervalSeconds\": 3600}"
```
Comment (omby8888) [inline integrations/custom/.port/spec.yaml]: maybe instead of this schema add a link to the docs so it will be accessible from the ui later on

File: integrations/custom/.port/spec.yaml [hunk 2]
```diff
@@ -156,6 +156,26 @@ configurations:
...
+    dependencies:
+      - on: authType
+        when:
+          equals: custom
+        then:
+          required: true
+  - name: customAuthResponse
+    required: false
+    type: object
+    description: "Custom authentication response usage configuration. Required when authType is 'custom'. Defines how to use auth response values in subsequent requests using template syntax {{.jq_path}}. Example: {\"headers\": {\"Authorization\": \"Bearer {{.access_token}}\"}, \"queryParams\": {\"api_key\": \"{{.access_token}}\"}, \"body\": {\"api_key\": \"{{.accessToken}}\"}}"
```
Comment (omby8888) [inline integrations/custom/.port/spec.yaml]: same

File: integrations/custom/http_server/auth/__init__.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+def get_auth_handler(
+    auth_type: str,
+    client: httpx.AsyncClient,
+    config: Dict[str, Any],
+) -> AuthHandler:
+    """Get the appropriate authentication handler"""
+    if auth_type == "custom":
+        return CustomAuthHandler(client, config)
+    handler_class = AUTH_HANDLERS.get(auth_type, NoAuth)
+    return handler_class(client, config)
```
Comment (omby8888) [inline integrations/custom/http_server/auth/__init__.py]: lets move that into dedicated factory file, and make the `AUTH_HANDLERS` private for that file only

File: integrations/custom/http_server/auth/simple.py
Comment (omby8888) [inline integrations/custom/http_server/auth/simple.py:1]: not sure the name is indicative here, also either put here the base class as well or split each Auth class to different file

File: integrations/custom/http_server/auth/simple.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class ApiKeyAuth(AuthHandler):
+    """API key authentication"""
+
+    def setup(self) -> None:
+        api_key = self.config.get("api_key")
+        key_header = self.config.get("api_key_header", "X-API-Key")
+        if api_key and key_header:
+            self.client.headers[key_header] = api_key
```
Comment (omby8888) [inline integrations/custom/http_server/auth/simple.py]:
```suggestion
        if api_key:
            key_header = self.config.get("api_key_header", "X-API-Key")
            self.client.headers[key_header] = api_key
```

File: integrations/custom/http_server/auth/custom_auth.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+        """Get cached evaluated templates."""
+        return (
+            self._evaluated_headers or {},
+            self._evaluated_query_params or {},
+            self._evaluated_body or {},
+        )
+
+    def update(
+        self,
+        hash_value: Optional[str],
```
Comment (omby8888) [inline integrations/custom/http_server/auth/custom_auth.py]:
```suggestion
        hash_value: str,
```
dont think this should be optional as u rely on either all values to exist in the `is_valid` check

File: integrations/custom/http_server/auth/custom_auth.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+    ):
+        self._auth_timestamp: Optional[float] = None
+        self._reauthenticate_interval = reauthenticate_interval
+        self._buffer_seconds = buffer_seconds
+
+    def record_authentication(self) -> None:
+        """Record that authentication happened."""
+        self._auth_timestamp = time.time()
+
+    def is_expired(self, has_auth_response: bool) -> bool:
```
Comment (omby8888) [inline integrations/custom/http_server/auth/custom_auth.py]:
```suggestion
    def should_reauthenticate(self, has_auth_response: bool) -> bool:
```
also `has_auth_response` is not required here

File: integrations/custom/http_server/auth/custom_auth.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+        else:
+            self._expiration_tracker = expiration_tracker
+
+    async def _perform_auth_request(self) -> None:
+        """Make authentication request asynchronously and store token (non-blocking)"""
+        if not self.custom_auth_request:
+            raise CustomAuthRequestError("customAuthRequest configuration is required")
+
+        logger.info("CustomAuth: Starting authentication")
+        logger.error("CustomAuth: running a new auth request")
```
Comment (omby8888) [inline integrations/custom/http_server/auth/custom_auth.py]: remove

File: integrations/custom/http_server/auth/custom_auth.py [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+
+        params = self.custom_auth_request.queryParams or {}
+        method = self.custom_auth_request.method
+        logger.debug(
+            f"CustomAuth: Making {method} request to {auth_url} for authentication"
+        )
+
+        async with httpx.AsyncClient(
+            verify=self.verify_ssl, timeout=DEFAULT_AUTH_TIMEOUT
+        ) as async_client:
```
Comment (omby8888) [inline integrations/custom/http_server/auth/custom_auth.py]: can use `port_ocean/utils/async_http.py`

File: integrations/custom/http_server/auth/custom_helpers/auth_flow.py [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+        if self.custom_auth_request.body:
+            json_data = self.custom_auth_request.body
+        elif self.custom_auth_request.bodyForm:
+            content = self.custom_auth_request.bodyForm
+
+        params = self.custom_auth_request.queryParams or {}
+        method = self.custom_auth_request.method
+        logger.debug(
+            f"CustomAuth: Making {method} request to {auth_url} for authentication"
+        )
```
Comment (omby8888) [inline integrations/custom/http_server/auth/custom_helpers/auth_flow.py:127]: can we split this to 2 different functions that each assemble the request, and in the ent sends param dict to the customer

File: integrations/custom/initialize_client.py [hunk 10]
```diff
@@ -62,3 +68,17 @@ def init_client() -> HttpServerClient:
...
+
+    Authentication is handled automatically by httpx.Auth when requests are made.
+
+    Returns:
+        HttpServerClient: The initialized client
+    """
+    global _client
+    if _client is None:
+        _client = init_client()
+    return _client
```
Comment (omby8888) [inline integrations/custom/initialize_client.py:84]: does this really functions as singleton in multiprocess?

File: integrations/custom/http_server/auth/custom/auth_flow.py [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+
+        # 2. Apply templates to the current request
+        authenticated_request = await self._override_request(request)
+
+        # 3. Send the request
+        response = yield authenticated_request
+
+        # 4. Reactive Auth Check (On status codes that require re-authentication)
+        if response.status_code in STATUS_CODES_TO_REAUTH:
+            async with self._reauth_lock_manager.lock:
```
Comment (omby8888) [inline integrations/custom/http_server/auth/custom/auth_flow.py]: this seem to return, maybe the lock logic should sit within _perform_auth_request

File: integrations/custom/http_server/auth/custom_helpers/auth_flow.py [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+        self, request: httpx.Request
+    ) -> AsyncGenerator[httpx.Request, httpx.Response]:
+        # 1. Proactive Auth Check (Before sending the request)
+        await self._ensure_authenticated()
+
+        # 2. Apply templates to the current request
+        authenticated_request = await self._override_request(request)
+
+        # 3. Send the request
+        response = yield authenticated_request
```
Comment (omby8888) [inline integrations/custom/http_server/auth/custom_helpers/auth_flow.py:218]: why the yield

File: integrations/custom/http_server/auth/custom_helpers/auth_flow.py
Comment (omby8888) [inline integrations/custom/http_server/auth/custom_helpers/auth_flow.py:1]: this file should be split to multiple files for readibilaty

File: integrations/custom/initialize_client.py (hunk 10 above)
Comment (shalev007) [reply integrations/custom/initialize_client.py:84]: it should act as a singleton since the kinds are not being processed in parallel, if so I think that simply using this in the @ocean.on_start sohuld resolve it but I'm not sure

File: integrations/custom/http_server/auth/__init__.py (hunk 3 above)
Comment (shalev007) [reply integrations/custom/http_server/auth/__init__.py]: Why?
Can't the __init__.py act as the factory file?

File: integrations/custom/http_server/auth/custom_auth.py (hunk 6 above)
Comment (shalev007) [reply integrations/custom/http_server/auth/custom_auth.py]: removed the `has_auth_response` but I refuse to rename the method

File: integrations/custom/http_server/auth/custom_helpers/auth_flow.py (hunk 12 above)
Comment (shalev007) [reply integrations/custom/http_server/auth/custom_helpers/auth_flow.py:218]: that yield hands the prepared request to httpx so the client can actually send it, and then httpx sends the received Response back into the async generator and assigns it to response.

File: integrations/custom/http_server/auth/custom/auth_flow.py (hunk 11 above)
Comment (shalev007) [reply integrations/custom/http_server/auth/custom/auth_flow.py]: what do you mean?

File: integrations/custom/http_server/auth/custom_auth.py (hunk 6 above)
Comment (omby8888) [reply integrations/custom/http_server/auth/custom_auth.py]: why?

File: integrations/custom/initialize_client.py (hunk 10 above)
Comment (omby8888) [reply integrations/custom/initialize_client.py:84]: kk i agree it should work then

File: integrations/custom/http_server/auth/__init__.py (hunk 3 above)
Comment (omby8888) [reply integrations/custom/http_server/auth/__init__.py]: imo it's not clear enough

File: integrations/custom/.port/spec.yaml [hunk 13]
```diff
@@ -156,6 +156,26 @@ configurations:
     type: string
     default: "10"
     description: "Maximum concurrent API requests (uses Ocean's built-in rate limiting and caching)"
+  - name: customAuthRequest
+    required: false
+    type: object
+    description: "Custom authentication request configuration. Required when authType is 'custom'. Defines how to make the authentication request. Example can be found in examples/config-custom-auth.env"
```
Comment (omby8888) [inline integrations/custom/.port/spec.yaml]: what about when it'll be shown in the UI? should have link to the docs here no?

File: integrations/custom/http_server/auth/custom_auth.py (hunk 6 above)
Comment (shalev007) [reply integrations/custom/http_server/auth/custom_auth.py]: because I think it's a more suitable name

File: integrations/custom/http_server/auth/__init__.py (hunk 3 above)
Comment (shalev007) [reply integrations/custom/http_server/auth/__init__.py]: I give up

File: integrations/custom/.port/spec.yaml (hunk 13 above)
Comment (shalev007) [reply integrations/custom/.port/spec.yaml]: I'm not sure if we should since the url might change in the future
